log.py

```diff
@@ -0,0 +1,165 @@
+    _last_applied: int = field(default=-1)
+    _entries: list[LogEntry] = field(default_factory=list)
+    _cur_state: dict[str, str] = field(default_factory=dict)
+    _commit_index: int = field(default=-1)
+    _apply_entries_task: asyncio.Task | None = field(default=None)
+
+    @classmethod
+    async def create(cls, path: str) -> Self:
+        ret: Self
+        file_exists: bool = os.path.isfile(path)
+        if file_exists:
+            fp = await aiofiles.open(path, "r+")
+            lines: list[str] = await fp.readlines()
+            entries: list[LogEntry] = []
+            cur_state: dict[str, str] = {}
+            for line in lines:
+                entry = json.loads(
+                    line, object_hook=JSONCodec.decode_log_entry
+                )
+                entries.append(entry)
+                cur_state[entry.key] = entry.value
+            last_applied = -1
+            if entries:
+                last_applied = entries[-1].index
+            ret = cls(path, fp, last_applied, entries, cur_state)
+        else:
+            fp = await aiofiles.open(ret._path, "w+")
+            ret = cls(path, fp)
+        return ret
+
+    def __del__(self: Self) -> None:
+        run_in_background(self._fp.close())
+
+    def is_empty(self: Self) -> bool:
+        return len(self._entries) == 0
+
+    def last_index_and_term(self: Self) -> tuple[int, int]:
+        last_index: int = -1
+        last_term: int = -1
+
+        if not self.is_empty():
+            last_entry: LogEntry = self._entries[-1]
+            last_index = last_entry.index
+            last_term = last_entry.term
+
+        return last_index, last_term
+
+    def term_for_index(self: Self, index: int) -> int:
+        return self._entries[index].term if index < len(self._entries) else -1
+
+    def get_entries_from(self: Self, start_index: int) -> list[LogEntry]:
+        return self._entries[start_index:]
+
+    def get_value(self: Self, key: str) -> str | None:
+        if key in self._cur_state:
+            return self._cur_state[key]
+        return None
+
+    def add_new_entry(self: Self, term: int, key: str, value: str) -> None:
+        index: int = self.last_index_and_term()[0] + 1
+        entry: LogEntry = LogEntry(
+            term=term, index=index, key=key, value=value
+        )
+        self._entries.append(entry)
+
+    def append_entries(self: Self, entries: list[LogEntry]) -> None:
+        # Need to overwrite any entries whose indices we already have
+        # entries for. Everything else can be appended.
+        for entry in entries:
+            # TODO: redo this logic when implementing log compaction
+            if entry.index < len(self._entries):
+                self._entries[entry.index] = entry
+            else:
+                assert entry.index == len(self._entries)
+                self._entries.append(entry)
+
+    def can_append_entries(
+        self: Self, append_entries_request: AppendEntriesRequest
```

Comment:
It feels a little bit wrong for your State Machine to see a raw RPC request type (especially since you only ever look at prev_log_index here).

Reply:
ohm you also look at prev_log_term, I can't read. Anyways, still probably better to just pass those two values in. And maybe worth minting a dataclass for `log_index_and_term`, since that seems to come up a bunch, but not a big deal.